File: apps/watch-wear/app/src/main/java/com/oneohone/wear/WatchPayload.kt
package com.oneohone.wear

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Compact binary wire format for wrist samples, byte-identical to the watchOS relay in
 * `apps/watch-ios/Sources/One01WatchCore/WatchPayload.swift`.
 *
 * Keeping one layout across both platforms means the phone side decodes wrist input the same way
 * regardless of which watch produced it, and `@101/adapter-watch` needs no per-platform branch.
 *
 * Layout (53 bytes, little-endian):
 * ```
 *  0      magic 0x31
 *  1      version
 *  2..3   sequence            u16
 *  4..7   milliseconds        u32   (since the relay started, not wall clock)
 *  8..23  orientation x,y,z,w f32 x4
 * 24..35  acceleration x,y,z  f32 x3 (g, gravity removed)
 * 36..47  rotationRate x,y,z  f32 x3 (degrees/second)
 * 48..51  crown travel        f32    (detents, monotonic)
 * 52      buttons bitfield    u8     (bit 0: tap)
 * ```
 *
 * Nothing here identifies the wearer or their health. 101 treats a watch as a motion controller
 * and nothing else.
 */
data class WatchPayload(
    val sequence: Int,
    val milliseconds: Long,
    val orientation: FloatArray,
    val acceleration: FloatArray,
    val rotationRate: FloatArray,
    val crown: Float,
    val tap: Boolean,
) {
    init {
        require(orientation.size == 4) { "orientation must be a quaternion" }
        require(acceleration.size == 3) { "acceleration must be a 3-vector" }
        require(rotationRate.size == 3) { "rotationRate must be a 3-vector" }
    }

    fun encode(): ByteArray {
        val buffer = ByteBuffer.allocate(BYTE_COUNT).order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(MAGIC)
        buffer.put(VERSION)
        buffer.putShort((sequence and 0xFFFF).toShort())
        buffer.putInt((milliseconds and 0xFFFFFFFFL).toInt())
        orientation.forEach(buffer::putFloat)
        acceleration.forEach(buffer::putFloat)
        rotationRate.forEach(buffer::putFloat)
        buffer.putFloat(crown)
        buffer.put(if (tap) 1 else 0)
        return buffer.array()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is WatchPayload) return false
        return sequence == other.sequence &&
            milliseconds == other.milliseconds &&
            orientation.contentEquals(other.orientation) &&
            acceleration.contentEquals(other.acceleration) &&
            rotationRate.contentEquals(other.rotationRate) &&
            crown == other.crown &&
            tap == other.tap
    }

    override fun hashCode(): Int {
        var result = sequence
        result = 31 * result + milliseconds.hashCode()
        result = 31 * result + orientation.contentHashCode()
        result = 31 * result + acceleration.contentHashCode()
        result = 31 * result + rotationRate.contentHashCode()
        result = 31 * result + crown.hashCode()
        result = 31 * result + tap.hashCode()
        return result
    }

    companion object {
        const val BYTE_COUNT = 53
        const val MAGIC: Byte = 0x31
        const val VERSION: Byte = 1

        /** Returns null rather than throwing on a short, misaligned or foreign packet. */
        fun decode(data: ByteArray): WatchPayload? {
            if (data.size != BYTE_COUNT) return null
            val buffer = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
            if (buffer.get() != MAGIC || buffer.get() != VERSION) return null

            val sequence = buffer.short.toInt() and 0xFFFF
            val milliseconds = buffer.int.toLong() and 0xFFFFFFFFL
            val orientation = FloatArray(4) { buffer.float }
            val acceleration = FloatArray(3) { buffer.float }
            val rotationRate = FloatArray(3) { buffer.float }
            val crown = buffer.float
            val tap = (buffer.get().toInt() and 0x01) == 0x01

            // A non-finite value would poison the motion pipeline downstream.
            val finite = orientation.all { it.isFinite() } &&
                acceleration.all { it.isFinite() } &&
                rotationRate.all { it.isFinite() } &&
                crown.isFinite()
            if (!finite) return null

            return WatchPayload(sequence, milliseconds, orientation, acceleration, rotationRate, crown, tap)
        }
    }
}

File: apps/watch-ios/Sources/One01WatchCore/WatchPayload.swift
import Foundation

/// Compact binary wire format for wrist samples.
///
/// A watch relays roughly 50 samples per second while a game is running, and every one of those
/// crosses a Bluetooth link with a small MTU and a real power cost. JSON would spend most of the
/// packet on key names that never change, so the realtime path uses a fixed little-endian layout
/// and reserves readable messages for the control path.
///
/// Layout (53 bytes):
/// ```
///  0      magic 0x31
///  1      version
///  2..3   sequence            u16
///  4..7   milliseconds        u32   (since the relay started, not wall clock)
///  8..23  orientation x,y,z,w f32 x4
/// 24..35  acceleration x,y,z  f32 x3 (g, gravity removed)
/// 36..47  rotationRate x,y,z  f32 x3 (degrees/second)
/// 48..51  crown travel        f32    (detents, monotonic)
/// 52      buttons bitfield    u8     (bit 0: tap)
/// ```
/// Nothing here identifies the wearer or their health. That is a deliberate limit, not an
/// omission: 101 treats a watch as a motion controller and nothing else.
public struct WatchPayload: Equatable, Sendable {
    public static let magic: UInt8 = 0x31
    public static let version: UInt8 = 1
    public static let byteCount = 53

    public var sequence: UInt16
    public var milliseconds: UInt32
    public var orientation: (x: Float, y: Float, z: Float, w: Float)
    public var acceleration: (x: Float, y: Float, z: Float)
    public var rotationRate: (x: Float, y: Float, z: Float)
    public var crown: Float
    public var tap: Bool

    public init(
        sequence: UInt16,
        milliseconds: UInt32,
        orientation: (x: Float, y: Float, z: Float, w: Float),
        acceleration: (x: Float, y: Float, z: Float),
        rotationRate: (x: Float, y: Float, z: Float),
        crown: Float,
        tap: Bool
    ) {
        self.sequence = sequence
        self.milliseconds = milliseconds
        self.orientation = orientation
        self.acceleration = acceleration
        self.rotationRate = rotationRate
        self.crown = crown
        self.tap = tap
    }

    public static func == (lhs: WatchPayload, rhs: WatchPayload) -> Bool {
        lhs.sequence == rhs.sequence
            && lhs.milliseconds == rhs.milliseconds
            && lhs.orientation == rhs.orientation
            && lhs.acceleration == rhs.acceleration
            && lhs.rotationRate == rhs.rotationRate
            && lhs.crown == rhs.crown
            && lhs.tap == rhs.tap
    }

    public func encoded() -> Data {
        var data = Data(capacity: Self.byteCount)
        data.append(Self.magic)
        data.append(Self.version)
        data.appendLittleEndian(sequence)
        data.appendLittleEndian(milliseconds)
        for value in [orientation.x, orientation.y, orientation.z, orientation.w] { data.appendLittleEndian(value) }
        for value in [acceleration.x, acceleration.y, acceleration.z] { data.appendLittleEndian(value) }
        for value in [rotationRate.x, rotationRate.y, rotationRate.z] { data.appendLittleEndian(value) }
        data.appendLittleEndian(crown)
        data.append(tap ? 1 : 0)
        return data
    }

    /// Returns nil rather than trapping on a short, misaligned or foreign packet. A relay must
    /// tolerate junk on the wire without taking the controller down.
    public static func decode(_ data: Data) -> WatchPayload? {
        guard data.count == byteCount else { return nil }
        let bytes = [UInt8](data)
        guard bytes[0] == magic, bytes[1] == version else { return nil }

        var offset = 2
        let sequence = bytes.littleEndianUInt16(at: &offset)
        let milliseconds = bytes.littleEndianUInt32(at: &offset)
        let qx = bytes.littleEndianFloat(at: &offset)
        let qy = bytes.littleEndianFloat(at: &offset)
        let qz = bytes.littleEndianFloat(at: &offset)
        let qw = bytes.littleEndianFloat(at: &offset)
        let ax = bytes.littleEndianFloat(at: &offset)
        let ay = bytes.littleEndianFloat(at: &offset)
        let az = bytes.littleEndianFloat(at: &offset)
        let gx = bytes.littleEndianFloat(at: &offset)
        let gy = bytes.littleEndianFloat(at: &offset)
        let gz = bytes.littleEndianFloat(at: &offset)
        let crown = bytes.littleEndianFloat(at: &offset)
        let tap = bytes[offset] & 0x01 == 0x01

        // A quaternion that is not finite would poison the motion pipeline downstream.
        for value in [qx, qy, qz, qw, ax, ay, az, gx, gy, gz, crown] where !value.isFinite {
            return nil
        }

        return WatchPayload(
            sequence: sequence,
            milliseconds: milliseconds,
            orientation: (qx, qy, qz, qw),
            acceleration: (ax, ay, az),
            rotationRate: (gx, gy, gz),
            crown: crown,
            tap: tap
        )
    }
}

private extension Data {
    mutating func appendLittleEndian(_ value: UInt16) {
        append(UInt8(value & 0xFF))
        append(UInt8((value >> 8) & 0xFF))
    }

    mutating func appendLittleEndian(_ value: UInt32) {
        for shift in stride(from: 0, to: 32, by: 8) { append(UInt8((value >> UInt32(shift)) & 0xFF)) }
    }

    mutating func appendLittleEndian(_ value: Float) {
        appendLittleEndian(value.bitPattern)
    }
}

private extension [UInt8] {
    func littleEndianUInt16(at offset: inout Int) -> UInt16 {
        defer { offset += 2 }
        return UInt16(self[offset]) | (UInt16(self[offset + 1]) << 8)
    }

    func littleEndianUInt32(at offset: inout Int) -> UInt32 {
        defer { offset += 4 }
        var value: UInt32 = 0
        for index in 0..<4 { value |= UInt32(self[offset + index]) << UInt32(index * 8) }
        return value
    }

    func littleEndianFloat(at offset: inout Int) -> Float {
        var local = offset
        let bits = littleEndianUInt32(at: &local)
        offset = local
        return Float(bitPattern: bits)
    }
}
